File: leetcode/39.py
# -*- coding: utf-8 -*-
# @Author: gunjianpan
# @Date:   2020-07-31 22:13:55
# @Last Modified by:   gunjianpan
# @Last Modified time: 2020-07-31 22:14:44

"""
39. Combination Sum Medium
Given a set of candidate numbers (candidates) (without duplicates) and a target number (target), find all unique combinations in candidates where the candidate numbers sums to target.

The same repeated number may be chosen from candidates unlimited number of times.

Note:

All numbers (including target) will be positive integers.
The solution set must not contain duplicate combinations.
Example 1:

Input: candidates = [2,3,6,7], target = 7,
A solution set is:
[
  [7],
  [2,2,3]
]
Example 2:

Input: candidates = [2,3,5], target = 8,
A solution set is:
[
  [2,2,2,2],
  [2,3,3],
  [3,5]
]

Constraints:

1 <= candidates.length <= 30
1 <= candidates[i] <= 200
Each element of candidate is unique.
1 <= target <= 500
Accepted 549,965 Submissions 983,999
"""


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        def dfs(idx, path, t):
            if not t:
                res.append(path.copy())
                return
            for ii in range(idx, N):
                tmp = t - candidates[ii]
                if tmp < 0:
                    break
                path.append(candidates[ii])
                dfs(ii, path, tmp)
                path.pop()

        N = len(candidates)
        if not N:
            return []
        candidates.sort()
        path, res = [], []
        dfs(0, path, target)
        return res

File: leetcode/39.java
class Solution {

  public void dfs(int[] s, int t, List<List<Integer>> res, List<Integer> temp,
                  int begin) {
    if (t == 0) {
      // System.out.print(temp);
      // System.out.print(t);
      System.out.print(" ");
      res.add(new ArrayList<>(temp));
      // System.out.print(temp);
      temp.clear();
      return;

    } else if (t < s[begin]) {
      temp.remove(temp.size() - 1);
      //  System.out.print(t);
      //  System.out.print(s[begin]);
      // System.out.print(temp);
      return;
    }

    // begin 防止查到前面去
    for (int i = begin; i < s.length; i++) {
      if (s[i] > t)
        break;
      if (s[i] <= t) {
        temp.add(s[i]);
        System.out.print(temp);
        System.out.print(t);
        // int[] b = Arrays.copyOfRange(s, i, s.length);
        dfs(s, t - s[i], res, temp, i);
      }
    }
    // return res;
  }
  public List<List<Integer>> combinationSum(int[] candidates, int target) {
    Arrays.sort(candidates);
    List<List<Integer>> res = new ArrayList<List<Integer>>();
    List<Integer> temp = new ArrayList<Integer>();
    dfs(candidates, target, res, temp, 0);

    return res;
  }
}